### ISSUE TITLE
Model roads as splines

### ISSUE THREAD

**Issue by martin-ez** [martin-ez/project-titan#4]
Roughly following [this suggestion](https://www.reddit.com/r/gamedev/comments/7vy53j/comment/dtw23bz/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).

> Your road network should be a series of interconnected splines (the connections being junctions)
> Each lane is, in effect it's own road, assuming you don't want cars making crazy U turns.
> Split each spline into a series of smaller road segments - try and keep these roughly the same length, as it'll simplify things greatly
> Each road segments is connected to the next one on the direction of travel. You can also connect a segment to the equivalent next segment in adjacent lanes, to let your routefinding handle lane changes.
> Each segment has a hard limit of how many vehicles can accommodate. A new vehicle can't enter until there's room for it. You could consider bigger vehicles to just be two or three cars together.
> Each car just has a float value from 0..1 indicating how far along its segment it is. You can calculate it's actual world position for rendering easily using the spline for the segment.
> The segment maintains a queue structure containing the vehicles, and is responsible for popping a vehicle off and pushing it to the next segment. The segment has a hard and soft speed limit - hard is physical maximum (for enforced slow traffic) - soft is the recommended maximum (so you can allow your drivers to break the speed limit if you wish)
> A vehicle can't overtake within a segment, so it can't go any faster than the vehicle ahead of it in the segment. To overtake, it needs to change lanes.
> Routefinding is just a*. You can have different weightings for different AI personalities. The ones that ignore traffic levels, and just stick to the most obvious route, the ones that avoid main roads and do the rat run down residential back streets. The one that always changes lane no matter what.
> Your points of interest again are just 0..1 positions on a particular road segment. I doubt your simulation will really matter if the car actually drives up to the actual location, rather than just stopping in the general vicinity, and turning into a parked car (or disappearing)
> If you do need parking, then just have a property per road segment that indicates how many vehicles can park there. Parking lots are just segments with 100 spaces.
> Random incidents are easy again, you just pick a segment, and set a flag that prevents vehicles leaving, or modifies the speed limit for the segment.
> Stoplights - again, a vehicle can't leave the segment if the signal is red.

**Comment by martin-ez** [comment]
#40 read 0.19's self-referential relationships as the thing that models a road. That is not quite what they are: `allow_self_referential` only lifts the ban on a relationship pointing at *its own entity*, which for a road would be a segment that is its own successor — a case worth allowing eventually for a loop of one, but not the feature this issue turns on.

The machinery this issue actually wants arrived in 0.16 and has been there through the migration: `#[relationship]` and `#[relationship_target]` keep both ends of a link in sync automatically and survive a despawn on either side. A `NextSegment(Entity)` on one segment with a `PreviousSegments` collection on the other is then a derive rather than a pair of systems that have to agree, and junctions fall out of the target being a collection rather than a single entity.

What 0.19 adds on top is real but small: relationships can be written inline in `bsn!`, and observers take `.run_if()` like ordinary systems.

Assessed under #40.

**Comment by martin-ez** [comment]
Narrowed to the spline and its segments. The body is a pasted source, not a specification, and it was carrying at least six separable pieces — most now filed against it.

What stays here: a road is a spline, split into segments of roughly equal length, each connected to the next in the direction of travel. That is the data model and nothing else.

What left, and where:

- #67 how lanes and direction work — a decision this cannot be built without, so it blocks this issue rather than following it
- #68 junctions, where splines connect
- #69 drawing a road with the road tool
- #70 the rover entity and its position along a segment
- #71 how an endpoint attaches to the network
- #73 a speed limit per segment
- #74 changing lanes to pass, conditional on #67
- #75 holding rovers at a junction with a signal

Segment capacity and the queue behind it stay with #8, which is the issue about roads being busy, not this one.

Written off from the source rather than filed: parking (a logistics fleet unloads and leaves, so a parked rover is a despawned one), random incidents (a progression mechanic at best, and nothing to attach it to yet), and driver personalities weighting routes differently (#6 can take a weight, but a fleet the player owns has no reason to disagree with itself).

### Done when

A road is a spline the player drew, split into segments of roughly even length that each know the next in the direction of travel, and a position along a segment converts to a world position through the spline rather than being stored as one.

**Comment by martin-ez** [comment]
#67 decided the lane model, and this is the segment definition it gives you. A road carries one lane in each direction, and `docs/road_network.md` records the decision and what it was chosen over.

- A road is what the player draws. A lane is a directed traversal of it. A segment belongs to exactly one lane.
- A segment links to the next segment in its own lane and to nothing in the other one. The source's adjacent-lane edge is deliberately dropped: there is no lane changing and no overtaking anywhere in the network.
- Every road has exactly two lanes, created together and removed together — #62 removes both or neither.
- Capacity and the queue are per segment, and therefore per lane. The two directions of a road never contend for the same room.
- The two lanes join at each end of a road, so a dead-end spur is drivable: a rover reaches the end and returns on the opposing lane. That is a property of the model, not a U-turn mechanism, and nothing else needs to own it.
- Junctions connect lane ends to lane starts, so a turn that crosses the opposing lane is #68's problem rather than a segment's.
- An endpoint attaches to a position on a lane's segment, so a building #71 connects is reachable from both directions.

When the road plugin lands, carry the decision into `RoadPlugin`'s doc comment — `SimulationPlugin` in `src/simulation.rs` is the precedent for a plugin doc holding the decision behind its design — and delete `docs/road_network.md` in the same change. It only lives outside the code because there is no plugin to hang it on yet.

**Comment by martin-ez** [comment]
Claimed by `feat/4-model-roads-as-splines` via `scripts/track.sh claim`.

**Comment by martin-ez** [comment]
Built by `feat/4-model-roads-as-splines`, waiting on a human merge of #90.

**Comment by martin-ez** [comment]
Superseded by #93. #92 decided that a road is a chain of circular arcs between nodes on the lattice the tile centres and their vertices form together, not a Catmull-Rom spline through the tiles it was drawn through. The source this issue quotes still holds everywhere else — a segment chain, a rover as a float along its segment, points of interest as positions on a segment, routing as A* — and #93, #70, #71 and #6 carry those forward unchanged. What went is the spline and the traced tile path. Invariants 3 and 6 in AGENTS.md now state the replacement.
